axelrod/graph.py

```diff
@@ -0,0 +1,126 @@
+"""
+Weighted undirected sparse graphs.
+
+Original source:
+https://github.com/marcharper/stationary/blob/master/stationary/utils/graph.py
+"""
+
+from collections import defaultdict
+
+
+class Graph(object):
+    """Weighted and directed graph object intended for the graph associated to a
+    Markov process. Gives easy access to the neighbors of a particular state
+    needed for various calculations.
+
+    Vertices can be any hashable / immutable python object. Initialize with a
+    list of edges:
+        [[node1, node2, weights], ...]
+    Weights can be omitted for an undirected graph.
+
+    For efficiency, neighbors are cached in dictionaries.
+    """
+
+    def __init__(self, edges=None, directed=False):
+        self.out_mapping = defaultdict(lambda: defaultdict(float))
+        self.in_mapping = defaultdict(lambda: defaultdict(float))
+        self.directed = directed
+        if edges:
+            self.add_edges(edges)
+
+    def add_vertex(self, label):
+        self._vertices.add(label)
+
+    def add_edge(self, source, target, weight=1.):
+        self.out_mapping[source][target] = weight
+        self.in_mapping[target][source] = weight
+        if not self.directed:
+            self.out_mapping[target][source] = weight
+            self.in_mapping[source][target] = weight
+
+    def add_edges(self, edges):
+        try:
+            for source, target, weight in edges:
+                self.add_edge(source, target, weight)
+        except ValueError:
+            for source, target in edges:
+                self.add_edge(source, target, 1.0)
+
+    def vertices(self):
+        """Returns the set of vertices of the graph."""
+        return list(self.out_mapping.keys())
+
+    def out_dict(self, source):
+        """Returns a dictionary of the outgoing edges of source with weights."""
+        return self.out_mapping[source]
+
+    def out_vertices(self, source):
```

Comment:
We may want to allow the graphs to be directed, see e.g. [here](http://www.usma.edu/nsc/siteassets/sitepages/publications/a%20novel%20analytical%20method%20for%20evolutionary%20graph.pdf).